Merge receipt into bundle (part 2)

liuchengxu opened the pull request:
Part 2 of #809, the substantial change in this PR is changing the receipt field in the Bundle structure from `bundle.receipt` to `bundle.receipts` so that the missing receipts can be submitted sooner in a batch. The validation of batched receipts is still based on the single receipt validation, one important thing is that we need to ensure the receipts are ordered properly. ~~It's supposed to write some tests, but I'd love to postpone it a little bit after the planned new features are finished.~~(basic tests in pallet-executor are fixed) Let me know if you have any confusion there.

The receipts validation on the pallet-executor has been done, but the validation of receipts from the gossiped bundle in x-net is left with a TODO since the bundle gossip is disabled for now anyway.                                                        

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](../CONTRIBUTING.md)

liuchengxu committed:
Prepare submitting the missing receipts

Move the code that supports submitting the missing receipts to the new place.

liuchengxu committed:
Update `bundle.receipt` to `bundle.receipts` to submit multiple missing receipts if any

liuchengxu committed:
Remove obsolete code

nazar-pc commented on crates/pallet-executor/src/lib.rs:
```diff
@@ -371,22 +376,30 @@ mod pallet {
                     let builder = ValidTransaction::with_tag_prefix("SubspaceSubmitBundle")
                         .priority(TransactionPriority::MAX)
-                        .and_provides(primary_number)
+                        .and_provides(first_primary_number)
```
This doesn't seem right. It should provide something other bundles would require. So maybe it should provide all the primary numbers contained in receipts instead and require the number right before the first one?

liuchengxu replied:
provides and requires should match each other, if you provide multiple numbers but require one number, they'll never match.

nazar-pc replied:
I meant calling `and_provides` multiple times, so that it provides multiple things, my assumption was that later transactions can depend on any of those things.

Either way I believe I found another issue here: it is broken if there is more than one ER because it'll "provide" the first block primary block here, while next bundle's ER with "require" the last primary block, so they will never actually match each other.

liuchengxu replied:
You can provide multiple things, but they are effectively the same because the receipt must be accepted consecutively, only the first primary block number matters and once it's included in some block, all the receipts will be stored on the primary chain.

>while next bundle's ER with "require" the last primary block, so they will never actually match each other.

Not sure I fully understand this. The next bundle's ER is always tied to the best receipt on the primary chain(starting from the next of the best primary receipt) and unrelated to the receipts in some tx which is still in the tx pool. If the best receipt on the primary chain is the same, the next bundle's "require" is the same with the previous bundle tx.

nazar-pc replied:
You have `and_requires` below that if submitted out of order will wait for `first_primary_number - One::one()` to arrive first. Now imagine that `first_primary_number - One::one()` is actually a second ER in the bundle. Since you only pick and provide the first primary number from the bundle, there may be no transaction that will provide the number expected.

liuchengxu replied:
That makes sense, it's a valid edge case, and now I also agree with providing multiple receipt numbers is inherently more correct.

nazar-pc commented on crates/pallet-executor/src/lib.rs:
```diff
@@ -371,22 +376,30 @@ mod pallet {
                     // primary_number is ensured to be larger than the best execution chain chain
                     // number above.
                     //
                     // No requires if it's the next expected execution chain number.
```
Why not though? It shouldn't hurt, the only exception would be the very first ER.

liuchengxu replied:
No requires means it can be put into the ready queue directly once validated, otherwise it will be put into the future queue in the tx pool. If it's the next expected one, then it should be just in the ready queue not in the future queue waiting for the missing tag to be satisfied.

nazar-pc replied:
Wouldn't the condition be satisfied immediately if you did actually have previous ER though?

liuchengxu replied:
Nope. The tags matching are entirely in the tx pool, it won't check if the previous ER is already included on chain. If you add a `requires` here, it would just wait until an extrinsic providing this ER is received, but it will never be satisfied as that extrinsic has been consumed and the tx pool won't see it again.

liuchengxu committed:
Rework pallet-executor tests

- Fix the existing tests
- Add a new test case in which multiple receipts are included in a bundle.

vedhavyas reviewed (approved):
Skipped tests
LGTM

liuchengxu committed:
Merge branch 'main' into merge-receipt-into-bundle-part-2

nazar-pc reviewed (commented):
Sorry, apparently I left comments, but forgot to submit review :disappointed:
